digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/dataimport/CsvDataImporter.scala

```diff
@@ -110,26 +119,73 @@ class TrafficSignCsvImporter extends CsvDataImporterOperations {
       (List(parameterName), Nil)
     }
   }
 
-  def tryToInt(propertyValue: String ) = {
+  def getRightRoadLinkUsingBearing(assetBearing: Option[Int], assetCoordinates: Point): (Seq[VVHRoadlink], Seq[RoadLink]) = {
+    val toleranceInDegrees = 25
+    val roadLinks = roadLinkService.getClosestRoadlinkForCarTrafficFromVVH(userProvider.getCurrentUser(), assetCoordinates)
+
+    assetBearing match {
+      case Some(aBearing) if roadLinks.nonEmpty =>
+        val enrichedRoadLinks = roadLinkService.enrichRoadLinksFromVVH(roadLinks)
+
+        val filteredEnrichedRoadLinks =
+          enrichedRoadLinks.filter { roadLink =>
+            val aTrafficDirection = TrafficDirection.apply(trafficSignService.getTrafficSignValidityDirection(assetCoordinates, roadLink.geometry))
+
+            val mValue = GeometryUtils.calculateLinearReferenceFromPoint(assetCoordinates, roadLink.geometry)
+            val roadLinkBearing = GeometryUtils.calculateBearing(roadLink.geometry, Some(mValue), Some(roadLink.length) )
+
+            if (roadLink.trafficDirection == TrafficDirection.BothDirections) {
+              val reverseRoadLinkBearing =
+                if (roadLinkBearing - 180 < 0) {
+                  roadLinkBearing + 180
+                } else {
+                  roadLinkBearing - 180
+                }
+
+              Math.abs(aBearing - roadLinkBearing) <= toleranceInDegrees ||
+                Math.abs(aBearing - reverseRoadLinkBearing) <= toleranceInDegrees
+            } else {
+              Math.abs(aBearing - roadLinkBearing) <= toleranceInDegrees &&
+                (roadLink.trafficDirection == aTrafficDirection || aTrafficDirection == TrafficDirection.BothDirections)
+            }
+          }
+
+        if (filteredEnrichedRoadLinks.nonEmpty) {
+          (roadLinks.filter { rl => filteredEnrichedRoadLinks.exists(_.linkId == rl.linkId) }, filteredEnrichedRoadLinks)
+        } else
+          (roadLinks, Seq())
+
+      case _ =>
+        (roadLinks, Seq())
+    }
+  }
+
+  def tryToInt(propertyValue: String ) : Option[Int] = {
     Try(propertyValue.toInt).toOption
   }
 
-  private def verifyData(parsedRow: CsvAssetRow): (List[String], Seq[VVHRoadlink]) = {
-    val lon = getPropertyValue(parsedRow, "lon").asInstanceOf[BigDecimal].toLong
-    val lat = getPropertyValue(parsedRow, "lat").asInstanceOf[BigDecimal].toLong
-    val roadLinks = roadLinkService.getClosestRoadlinkForCarTrafficFromVVH(userProvider.getCurrentUser(), Point(lon, lat))
+  private def verifyData(parsedRow: CsvAssetRow): ParsedCsv = {
+    val optLon = getPropertyValueOption(parsedRow, "lon").asInstanceOf[Option[BigDecimal]]
+    val optLat = getPropertyValueOption(parsedRow, "lat").asInstanceOf[Option[BigDecimal]]
+    val bearing = tryToInt(getPropertyValue(parsedRow, "bearing").toString)
 
-    if(roadLinks.isEmpty) {
-      (List(s"Try to create in an unauthorized Municipality"), Seq())
-    } else
-      (List(), roadLinks)
+    (optLon, optLat) match {
```

Comment:
the coordinates are not optional so we can avoid matching here

Reply:
here the coordinates should be get as an optional, the goal is check all possible errors, and if doesn't exist a coordinate the process will abend

because of that
 **case _ =>
        (Nil, Nil) //That condition is already checked on assetRowToProperties**